### PR TITLE
Add mate::Promise

### PR DESCRIPTION
**Pull request by MarshallOfSound**
This is just so we can start looking at native promisification in Electron 👍

### CODE REVIEW

File: native_mate/dictionary.cc [hunk 1]
```diff
@@ -20,7 +20,7 @@ Dictionary::~Dictionary() {
 }
 
 Dictionary Dictionary::CreateEmpty(v8::Isolate* isolate) {
-  return Dictionary(isolate, v8::Object::New(isolate));;
```
Comment (MarshallOfSound) [inline native_mate/dictionary.cc:23]: Yeah this bugged me 😆

File: native_mate/promise.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+ private:
+  v8::Local<v8::Promise::Resolver> resolver_;
+};
+
+template<>
+struct Converter<Promise> {
+  static v8::Local<v8::Value> ToV8(v8::Isolate* isolate,
+                                    Promise val);
+  // TODO(MarshallOfSound): Implement FromV8 to allow promise chaining
```
Comment (MarshallOfSound) [inline native_mate/promise.h:48]: We don't have a use case for this yet and it's actually quite a hard thing to do, so gonna leave it out till we need to take in promises (don't see why we would yet)

File: native_mate/promise.h [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+namespace mate {
+
+class Promise {
+ public:
+  Promise();
+  Promise(v8::Isolate* isolate);
+  virtual ~Promise();
+
+  static Promise Create(v8::Isolate* isolate);
+  static Promise Create();
```
Comment (zcbenz) [inline native_mate/promise.h:19]: The `Create` methods are unnecessary.

**Review by zcbenz — CHANGES_REQUESTED** [review]

File: native_mate/promise.cc [hunk 4]
```diff
@@ -0,0 +1,45 @@
+// Copyright (c) 2018 GitHub, Inc.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include "native_mate/promise.h"
+
+namespace mate {
+
+Promise::Promise()
+    : isolate_(NULL) {
```
Comment (alexeykuzmin) [inline native_mate/promise.cc:10]: `nullptr` maybe?

File: native_mate/promise.cc [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  v8::Local<v8::Value> error = v8::Exception::Error(error_message);
+  resolver_->Reject(mate::ConvertToV8(isolate(), error));
+}
+
+v8::Local<v8::Object> Promise::GetHandle() const {
+  return resolver_->GetPromise();
+}
+
+v8::Local<v8::Value> Converter<Promise>::ToV8(v8::Isolate* isolate,
+                                                  Promise val) {
```
Comment (alexeykuzmin) [inline native_mate/promise.cc:41]: Indentation is a bit off here.

File: native_mate_files.gypi [hunk 6]
```diff
@@ -22,6 +22,8 @@
       'native_mate/wrappable.cc',
       'native_mate/wrappable.h',
       'native_mate/wrappable_base.h',
+      'native_mate/promise.h',
+      'native_mate/promise.cc',
```
Comment (alexeykuzmin) [inline native_mate_files.gypi:26]: The list seems to be sorted alphabetically.

**Comment by YurySolovyov** [comment]
Closes/Refs #20 ?

**Comment by MarshallOfSound** [comment]
@alexeykuzmin @zcbenz Updated as per feedback 👍

**Review by alexeykuzmin — APPROVED** [review]

**Pull request merged**